src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -1820,6 +1911,244 @@ public string DeleteCounter(DocumentsOperationContext context, string documentId
+                        {
+                            table.DeleteByKey(counterGroupKey);
+                        }
+                    }
+                    else
+                    {
+                        using var scope = Slice.From(context.Allocator, existing.Read((int)CountersTable.CounterKey, out var size), size, out var counterGroupKey);
+
+                        var prop = new BlittableJsonReaderObject.PropertyDetails();
+                        names.GetPropertyByIndex(0, ref prop);
+                        if (prop.Name.Equals(item.Name) && counterGroupKey.Size == counterKeySlice.Size)
+                        {
+                            // we need to change the counter group key and delete the old one
+                            table.DeleteByKey(counterGroupKey);
+
+                            names.GetPropertyByIndex(1, ref prop);
+                            using var newScope = context.Allocator.Allocate(CounterKeysSlice.Size + 1 /* separator */  + 1 /* replace the current name with next one in counters group */, out ByteString newCounterKey);
+                            documentKeyPrefix.CopyTo(newCounterKey.Ptr);
+                            Memory.Copy(newCounterKey.Ptr + documentKeyPrefix.Size, prop.Name.Buffer, prop.Name.Size);
+                            Slice.From(context.Allocator, newCounterKey.Ptr, documentKeyPrefix.Size + prop.Name.Size, out counterGroupKey);
+                        }
+
+                        counterValues.Modifications ??= new DynamicJsonValue(counterValues);
+                        counterValues.Modifications.Remove(item.Name);
+
+                        names.Modifications ??= new DynamicJsonValue(names);
+                        names.Modifications.Remove(item.Name);
+
+                        data.Modifications = new DynamicJsonValue(data);
+                        using (var old = data)
+                        {
+                            data = context.ReadObject(data, item.DocumentId, BlittableJsonDocumentBuilder.UsageMode.ToDisk);
+                        }
+
+                        var newEtag = _documentsStorage.GenerateNextEtag();
+                        var newChangeVector = _documentsStorage.GetNewChangeVector(context, newEtag);
+
+                        using (Slice.From(context.Allocator, newChangeVector, out var cv))
+                        using (DocumentIdWorker.GetStringPreserveCase(context, collectionName.Name, out Slice collectionSlice))
+                        using (table.Allocate(out TableValueBuilder tvb))
+                        {
+                            tvb.Add(counterGroupKey);
+                            tvb.Add(Bits.SwapBytes(newEtag));
+                            tvb.Add(cv);
+                            tvb.Add(data.BasePointer, data.Size);
+                            tvb.Add(collectionSlice);
+                            tvb.Add(context.GetTransactionMarker());
+
+                            table.Set(tvb);
+                        }
```

Comment:
we are modifying a `CounterGroup` and calling `table.Set(tvb)` on each counter-deletion. 
so if some of the counters-to-delete are in the same `CounterGroup` - we will modify the same group multiple times.
we can probably optimize this to modify each `CounterGroup` just once (in a similar way to what we do in [incoming counter replication](https://github.com/ravendb/ravendb/blob/v5.3/src/Raven.Server/Documents/CountersStorage.cs#L1017))

Reply:
As agreed with @karmeli87 & @aviv86 - this is an optimization, so for now, we'll keep it as is. In the future, we'll maybe change it (if we notice some performance issues or if we like :-))